disk/src/com/treode/disk/edit/PageWriter.scala

```diff
@@ -15,32 +14,75 @@
  */
 
 package com.treode.disk.edit
-
+import com.treode.async.{Async, Callback, Scheduler}, Async.supply, Callback.ignore
 import com.treode.async.io.File
 import com.treode.async.Async
 import com.treode.buffer.PagedBuffer
+import scala.collection.mutable.UnrolledBuffer
+import com.treode.async.implicits._
+import com.treode.disk.Position
+import com.treode.async.{Async, Callback, Scheduler}, Async.async
+import com.treode.disk.PickledPage
 
-private class PageWriter (
-  val file: File
-) {
 
+
+/*
+    PageWriters get items to write from a PageDispatcher and
+    write them to a file.
+ */
+private class PageWriter(dsp: PageDispatcher,val file: File)
+ {
   val bits = 10
   val buffer = PagedBuffer (bits)
-  var pos : Long = 0
+  // we maintain an internal pointer into the file for where to write
+  // in order to preserve consistency between writes.
+  var pos : Long = buffer.writePos 
+  val dispatcher = dsp
+  
+
+  // we consistently check to see if we have anything to write.
+  listen() run (ignore)
+
 
-  /**
-   * Write `data` into the file asynchronously, using a write buffer. Returns
-   * the position of the writer and length written, if successful.
+  /*
+    This polls the attached PageDispatcher for items to write to the
+    attached file.
    */
-  def write (data: String) : Async [(Long, Int)] = {
-    buffer.writeString (data)
-    val len = buffer.writePos
+  def listen(): Async[Unit] =
     for {
-      _ <- file.flush (buffer, pos)
+      (_, dataBuffers) <- dispatcher.receive()
+      _ <- write(dataBuffers)
+    } yield {
+      listen() run (ignore)
+    }
+
+    /*
+      Write all of the PickledPages data in the UnrolledBuffer 
+      to the file, then return the position where the PickledPages were
+      written to the callback inside the PickledPage.   
+     */
+   def write(data: UnrolledBuffer[PickledPage]): Async [Unit] = {
+    var writePositions : Array[Position] = new Array[Position](0);
+    var beforeAnyWrites = pos
+    var i = 0; 
```

Comment:
oh sorry - it used to be used to index into writePositions, but then I just started appending ...